### ISSUE TITLE
Cannot backup production loads to AWS s3 buckets ("Broken pipe")

### ISSUE THREAD

**Issue by michaelsembwever** [thelastpickle/cassandra-medusa#20]
Uploading large SSTables to an s3 bucket often experiences `BrokenPipeError` failures, so in practice production backups to an s3 bucket will fail.

The Apache Libcloud library used has no retry policy when uploading chunks/mutiparts to an s3 bucket. In addition the Medusa code does not use the chunk/multipart upload libcloud api.

The fix for this is two part: 
1.  the Medusa code needs to use libcloud's api for multipart uploads, and deal with AWS's different approach to hashing multipart-uploaded files, and
2.  the libcloud library needs to have a retry policy added to it.

(1) is available in https://github.com/thelastpickle/cassandra-medusa/compare/alex/fix-s3-broken-pipes

(2) is available in https://github.com/apache/libcloud/compare/trunk...thelastpickle:retry-s3-multipart-uploads

**Comment by michaelsembwever** [comment]
fyi @Kami

**Comment by michaelsembwever** [comment]
After a number of experiments we have landed on that libcloud is not capable of reliably uploading files to an s3 bucket fast enough.

The use of libcloud has been replaced with direct calls to `aws s3 cp …` which performs well. It provides automatic multipart-ing, concurrency, and throttling. (An initial experiment demonstrated an upload rate 100x that of libcloud, which was dangerous to the server at hand, so throttling is crucial.) 

See https://github.com/thelastpickle/cassandra-medusa/pull/22 for the work-in-progress
